src/psyclone/domain/nemo/transformations/nemo_loop_fuse.py

```diff
@@ -0,0 +1,237 @@
+
+        :param node1: the first Node that is being checked.
+        :type node1: :py:class:`psyclone.psyir.nodes.Node`
+        :param node2: the second Node that is being checked.
+        :type node2: :py:class:`psyclone.psyir.nodes.Node`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        :raises TransformationError: if the lower or upper loop boundaries \
+            are not the same.
+        :raises TransformationError: if the loop step size is not the same.
+        :raises TransformationError: if the loop variables are not the same.
+        '''
+
+        # First check constraints on Nodes in the node_list inherited from
+        # the parent classes (ExtractTrans and RegionTrans)
+        super(NemoLoopFuseTrans, self).validate(node1, node2, options)
+
+        if not node1.start_expr.math_equal(node2.start_expr):
+            raise TransformationError("Lower loop bounds must be identical, "
+                                      "but are '{0}'' and '{1}'"
+                                      .format(node1.start_expr,
+                                              node2.start_expr))
+        if not node1.stop_expr.math_equal(node2.stop_expr):
+            raise TransformationError("Upper loop bounds must be identical, "
+                                      "but are '{0}'' and '{1}'"
+                                      .format(node1.stop_expr,
+                                              node2.stop_expr))
+        if not node1.step_expr.math_equal(node2.step_expr):
+            raise TransformationError("Step size in loops must be identical, "
+                                      "but are '{0}'' and '{1}'"
+                                      .format(node1.step_expr,
+                                              node2.step_expr))
+        loop_var1 = node1.variable.name
+        loop_var2 = node2.variable.name
+
+        if loop_var1 != loop_var2:
+            raise TransformationError("Loop variables must be the same, "
+                                      "but are '{0}' and '{1}'".
+                                      format(loop_var1, loop_var2))
+        vars1 = VariablesAccessInfo(node1)
+        vars2 = VariablesAccessInfo(node2)
+
+        # Get all variables that occur in both loops. A variable
+        # that is only in one loop is not affected by fusion.
+        all_vars = set(vars1).intersection(vars2)
+        symbol_table = node1.scope.symbol_table
+
+        for var_name in all_vars:
+            # Ignore the loop variable
+            if var_name == loop_var1:
+                continue
+            var_info1 = vars1[var_name]
+            var_info2 = vars2[var_name]
+
+            # Variables that are only read in both loops can always be fused
+            if var_info1.is_read_only() and var_info2.is_read_only():
+                continue
+
+            if var_name in symbol_table:
```

Comment:
I have re-implemented this, I believe the 'in' operation would have caused problems (since I believe that would only look at the specified symbol table, not the hierarchy of them). Please check if I did that correctly.

Reply:
Looks good to me. I would recommend having a test with a suitable variable defined in a module and then accessed in a subroutine - I haven't got as far as looking at the tests yet so you may have already done this.

Reply:
I would actually say that this needs to be tested in the symbol table code (or as part of #1213), imho it doesn't make any sense to add this when evaluating loop fusion.